core/src/main/java/jenkins/install/StartupUtil.java

```diff
@@ -0,0 +1,159 @@
+import java.io.File;
+import java.io.IOException;
+import java.util.logging.Logger;
+
+import static java.util.logging.Level.SEVERE;
+
+/**
+ * Jenkins startup utilities.
+ *
+ * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
+ * @since FIXME
+ */
+public class StartupUtil {
+
+    private static final Logger LOGGER = Logger.getLogger(StartupUtil.class.getName());
+
+    private static final VersionNumber NEW_INSTALL_VERSION = new VersionNumber("1.0");
+    
+    /**
+     * Get the type of "startup" currently under way in Jenkins.
+     * @return The type of "startup" currently under way in Jenkins.
+     */
+    public static StartupType getStartupType() {
+        if (Functions.getIsUnitTest()) {
+            return StartupType.TEST;
+        }
+        
+        VersionNumber lastRunVersion = new VersionNumber(getLastExecVersion());
+
+        // Neither the top level config or the lastExecVersionFile have a version
+        // stored in them, which means it's a new install.
+        if (lastRunVersion.compareTo(NEW_INSTALL_VERSION) == 0) {
+            return StartupType.NEW;
+        }
+
+        // We have a last version.
+
+        VersionNumber currentRunVersion = new VersionNumber(getCurrentExecVersion());
+        if (lastRunVersion.isOlderThan(currentRunVersion)) {
+            return StartupType.UPGRADE;
+        } else if (lastRunVersion.isNewerThan(currentRunVersion)) {
+            return StartupType.DOWNGRADE;
+        } else {
+            // Last running version was the same as "this" running version.
+            return StartupType.RESTART;
+        }
+    }
+
+    /**
+     * Save the current Jenkins instance version as the last executed version.
+     * <p>
+     * This state information is required in order to determine whether or not the Jenkins instance
+     * is just restarting, or is being upgraded from an earlier version.
+     */
+    public static void saveLastExecVersion() {
+        if (Jenkins.VERSION.equals(Jenkins.UNCOMPUTED_VERSION)) {
+            // This should never happen!! Only adding this check in case someone moves the call to this method to the wrong place.
+            throw new IllegalStateException("Unexpected call to StartupUtil.saveLastExecVersion(). Jenkins.VERSION has not been initialized. Call computeVersion() first.");
+        }
+        saveLastExecVersion(Jenkins.VERSION);
+    }
+
+    /**
+     * Get the last saved Jenkins instance version.
+     * @return The last saved Jenkins instance version.
+     * @see #saveLastExecVersion()
+     */
+    public static @Nonnull String getLastExecVersion() {
+        File lastExecVersionFile = getLastExecVersionFile();
+        if (lastExecVersionFile.exists()) {
+            try {
+                return FileUtils.readFileToString(lastExecVersionFile);
+            } catch (IOException e) {
+                throw new IllegalStateException("Unexpected Error. Unable to read " + lastExecVersionFile.getAbsolutePath(), e);
```

Comment:
Perhaps; we do not really know. Are you positive you want to halt startup over this?